cloudinit/kindadapter.go

```diff
@@ -75,38 +76,36 @@ func Run(userData []byte, cmder exec.Cmder) ([]string, error) {
 	return lines, nil
 }
 
-// getCloudCongfigActionSlices parse the cloud config yaml into a slice of cloud config actions.
-// NB. it is necessary to parse manually because a cloud config is a map of unstructured elements;
-// using a standard yaml parser and make it UnMarshal to map[string]interface{} can't work because
-// [1] map does not guarantee the order of items, while the adapter should preserve the order of actions
-// [2] map does not allow to repeat the same action module in two point of the cloud init sequence, while
-// this might happen in real cloud inits
-func getCloudCongfigActionSlices(userData []byte) ([]cloudCongfigAction, error) {
-	var cloudConfigActionRegEx = regexp.MustCompile(`^[a-zA-Z_]*:`)
+// getActionSlices parses the cloud config yaml into a slice of actions to run.
+// Parsing manually is required because the order of the cloud config's actions must be maintained.
```

Comment:
This doesn't line up with my understanding of cloud-init. The order of the modules does not matter, and I'm not sure that cloud-init will honor a module specified multiple times, since it is parsing the file using the python yaml parser. There are configurable merging behaviors when specifying [multiple files](https://cloudinit.readthedocs.io/en/latest/topics/merging.html), but not within a single file afaict.

The order of modules run is determined by a combination of [boot stages](https://cloudinit.readthedocs.io/en/latest/topics/boot.html), on disk config, and the `cloud_init_modules`, `cloud_config_modules`, and `cloud_final_modules` keys in the cloud-config provided.

Reply:
ah hmm, ok...

I don't know how much "adapting" we want to do here then... I'm hoping cloud-init generated will be simple enough that this will be sufficient? We really do *not* want to be trying to rebuild cloud-init in any way on this provider.

Reply:
I wasn't meaning to imply that we should rebuild all that functionality, just that we do not need to worry about map ordering or running a module multiple times.

For ordering, we can just do all files prior to all run commands, since that is the default config used by cloud-init.

Reply:
ah ha! I will update the comment. 

The reality is that cloud-init is essentially a map[string]interface{} but the keys are not unique haha so we will have to parse manually anyway. i'm fairly sure since it's not actually a real list of items we can't get it into a []map[string]interface{}. I will update the comment to reflect my new understanding of reality.

Reply:
on further inspection i'm wrong and have simply been reading the comments too closely. 😂

Reply:
I'm going to write up an issue for improvement that can get rid of the custom parsing.